Scan BidiRTL possibility via SIMD

Constellation opened the pull request:
#### 3e50aeaaa02169629761b2f64041cedd72c4fedd
<pre>
Scan BidiRTL possibility via SIMD
<a href="https://bugs.webkit.org/show_bug.cgi?id=274705">https://bugs.webkit.org/show_bug.cgi?id=274705</a>
<a href="https://rdar.apple.com/128719350">rdar://128719350</a>

Reviewed by Sam Weinig.

Deploy optimization for BidiRTL possibility scanning via SIMD.
We clean up SIMDHelpers more to make code simpler

1. Add variadic bitOr / bitAnd / merge to simplify existing implementations.
2. Add all comparisons (lessThan, lessThanOrEqual, greaterThan, greaterThanOrEqual).
3. Make splat more easy-to-use form.
4. Add SIMD::stride helper inline variable to compute stride easily.

For existing implementations, I just deployed this new form. But in the furture, we would like
to extract common pattern and further simplifies the implementations. But for now, let&apos;s just
make BidiRTL scanning fast.

* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier&lt;CharType&gt;::append):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser&lt;CharType&gt;::Lexer::lexString):
* Source/WTF/wtf/SIMDHelpers.h:
(WTF::SIMD::splat8):
(WTF::SIMD::splat16):
(WTF::SIMD::splat32):
(WTF::SIMD::splat64):
(WTF::SIMD::splat):
(WTF::SIMD::merge2):
(WTF::SIMD::bitOr2):
(WTF::SIMD::bitAnd2):
(WTF::SIMD::merge):
(WTF::SIMD::bitOr):
(WTF::SIMD::bitAnd):
(WTF::SIMD::bitNot):
(WTF::SIMD::lessThanOrEqual):
(WTF::SIMD::greaterThan):
(WTF::SIMD::greaterThanOrEqual):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findImpl):
(WTF::charactersContain):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanAttributeValue):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::containsStrongDirectionalityText):

Canonical link: <a href="https://commits.webkit.org/279356@main">https://commits.webkit.org/279356@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7841a3fefefb234d272b8c855302919358347a60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53279 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5776 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56558 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55585 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39944 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3750 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43192 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/2615 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55377 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30729 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46010 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24322 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3334 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2161 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/46635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49429 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/3494 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58154 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52793 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28426 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3481 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50592 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/29643 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46223 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49915 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/30565 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/65097 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7828 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29404 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12403 "Found 6 new JSC stress test failures: stress/proxy-set-prototype-of.js.mini-mode, stress/spread-non-array.js.no-llint, wasm.yaml/wasm/function-tests/i32-const.js.wasm-eager, wasm.yaml/wasm/function-tests/invalid-duplicate-export.js.wasm-eager, wasm.yaml/wasm/references/is_null.js.wasm-eager, wasm.yaml/wasm/stress/wasm-to-wasm.js.wasm-eager (failure)") | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
```diff
@@ -483,7 +483,48 @@ bool TextUtil::containsStrongDirectionalityText(StringView text)
 
-    if (text.containsOnly<isNotBidiRTL>())
+    if (![&](auto span) ALWAYS_INLINE_LAMBDA {
+        using UnsignedType = std::make_unsigned_t<typename decltype(span)::value_type>;
```
This is all very cool, but also a bit challenging to understand. I think these SIMD uses could really benefit from a block comment at the their head explaining what the following code is intending to do (I also find ascii art helps for understanding SIMD as well).

Constellation replied:
Yeah, right now it exists on SIMDHelpers.h.

weinig replied:
I guess I meant something more along the lines of describing what the use of the SIMD operations is doing, rather than how operations work themselves. 

Take the scanText() function in the HTMLFastPathParser.  I was thinking something as simple as a comment saying something like:

"This function scans for the characters '<', '&', '\r' and '\0' in `m_parsingBuffer` using SIMD to perform the checks on strides of 16 characters at a time."

While not something that we would do in an ordinary code, I think the novelty of using SIMD makes a bit more explanation worthwhile.

weinig commented on Source/JavaScriptCore/runtime/JSONObject.cpp:
```diff
@@ -1063,13 +1063,13 @@ void FastStringifier<CharType>::append(JSValue value)
         auto charactersCopySameType = [&](auto span, auto* cursor) ALWAYS_INLINE_LAMBDA {
 #if (CPU(ARM64) || CPU(X86_64)) && COMPILER(CLANG)
-            constexpr size_t stride = 16 / sizeof(CharType);
+            constexpr size_t stride = SIMD::stride<CharType>;
```
Maybe with this new round of optimization, it will finally be the time to unify the JSC and WTF json parsers :).

Constellation replied:
I'm having a bit concern about it. JSC side JSON is super optimized with JSC object etc., so this is very hard to unify them without regressing JSC performance, and given that WTF side is not so sensitive to performance, I think rather separating them would be simpler.

weinig replied:
Yeah, that has been the concern all along :). I still believe there is a nice templated chunk that can be shared for the lexing/parsing, but it is certainly a reach.

weinig commented on Source/WTF/wtf/SIMDHelpers.h:
```diff
@@ -91,26 +136,113 @@ ALWAYS_INLINE void store(simde_uint64x2_t value, uint64_t* ptr)
 }
 
-ALWAYS_INLINE simde_uint8x16_t merge(simde_uint8x16_t accumulated, simde_uint8x16_t input)
+ALWAYS_INLINE simde_uint8x16_t merge2(simde_uint8x16_t accumulated, simde_uint8x16_t input)
```
Why are these ones just ALWAYS_INLINE, and not constexpr? Are the underlying intrinsics not constexpr?

Constellation replied:
Because simde functions are not constexpr.

weinig commented on Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
```diff
@@ -480,21 +480,21 @@ class HTMLFastPathParser {
-                const auto escapeMask = SIMD::splat(static_cast<UnsignedType>('&'));
-                const auto newlineMask = SIMD::splat(static_cast<UnsignedType>('\r'));
-                const auto zeroMask = SIMD::splat(static_cast<UnsignedType>(0));
+                const auto quoteMask = SIMD::splat<UnsignedType>('<');
```
Can these be `constexpr` like is done in the `LiteralParser` for splats?

Constellation replied:
Done.

weinig commented on Source/JavaScriptCore/runtime/LiteralParser.cpp:
```diff
@@ -869,18 +869,18 @@ ALWAYS_INLINE TokenType LiteralParser<CharType>::Lexer::lexString(LiteralParserT
             ([&]() ALWAYS_INLINE_LAMBDA {
 #if CPU(ARM64) || CPU(X86_64)
-                constexpr size_t stride = 16 / sizeof(CharType);
+                constexpr size_t stride = SIMD::stride<CharType>;
```
Other than the specifics of what characters are being searched for, this looks very close to same code as is used in the HTMLFastPath code below. 

Is there a common bit that could be pulled out? Something like the following (totally untested or compiled, just written here in GitHub).

```c++
template<typename CharType, CharType... Characters>
const CharType* searchFor(const CharType* ptr, const CharType* end)
{
    constexpr size_t stride = SIMD::stride<CharType>;
    using UnsignedType = std::make_unsigned_t<CharType>;
    if (static_cast<size_t>(end - ptr) >= stride) {
        auto match = [&](auto* cursor) ALWAYS_INLINE_LAMBDA {
            auto input = SIMD::load(bitwise_cast<const UnsignedType*>(cursor));
            auto mask = SIMD::bitOr(SIMD::equal(input, SIMD::splat<UnsignedType>(Characters))...);
            return SIMD::findFirstNonZeroIndex(mask);
        };

        for (; ptr + (stride - 1) < end; ptr += stride) {
            if (auto index = match(ptr)) {
                ptr += index.value();
                return ptr;
            }
        }
        if (ptr < end) {
            if (auto index = match(end - stride)) {
                ptr = end - stride + index.value();
                return ptr;
            }
            ptr = end;
        }
        return ptr;
    }

    for (; ptr != end; ++ptr) {
        auto character = *ptr;
        if (... || (character == characters))
            return ptr;
    }

    return ptr;
}
```

Constellation replied:
Right now, this patch is focusing on adding optimization. Cleaning up will be done in a separate change. We are still in an early phase and avoiding too-early-clean-up until we extract enough patterns of use of SIMD in our code. After we collected enough cases, we will do some clean up.

weinig replied:
Makes sense.

Constellation committed:
Scan BidiRTL possibility via SIMD
https://bugs.webkit.org/show_bug.cgi?id=274705
rdar://128719350

Reviewed by Sam Weinig.

Deploy optimization for BidiRTL possibility scanning via SIMD.
We clean up SIMDHelpers more to make code simpler

1. Add variadic bitOr / bitAnd / merge to simplify existing implementations.
2. Add all comparisons (lessThan, lessThanOrEqual, greaterThan, greaterThanOrEqual).
3. Make splat more easy-to-use form.
4. Add SIMD::stride helper inline variable to compute stride easily.

For existing implementations, I just deployed this new form. But in the furture, we would like
to extract common pattern and further simplifies the implementations. But for now, let's just
make BidiRTL scanning fast.

* Source/JavaScriptCore/runtime/JSONObject.cpp:
(JSC::FastStringifier<CharType>::append):
* Source/JavaScriptCore/runtime/LiteralParser.cpp:
(JSC::LiteralParser<CharType>::Lexer::lexString):
* Source/WTF/wtf/SIMDHelpers.h:
(WTF::SIMD::splat8):
(WTF::SIMD::splat16):
(WTF::SIMD::splat32):
(WTF::SIMD::splat64):
(WTF::SIMD::splat):
(WTF::SIMD::merge2):
(WTF::SIMD::bitOr2):
(WTF::SIMD::bitAnd2):
(WTF::SIMD::merge):
(WTF::SIMD::bitOr):
(WTF::SIMD::bitAnd):
(WTF::SIMD::bitNot):
(WTF::SIMD::lessThanOrEqual):
(WTF::SIMD::greaterThan):
(WTF::SIMD::greaterThanOrEqual):
* Source/WTF/wtf/text/StringCommon.h:
(WTF::findImpl):
(WTF::charactersContain):
* Source/WebCore/html/parser/HTMLDocumentParserFastPath.cpp:
(WebCore::HTMLFastPathParser::scanText):
(WebCore::HTMLFastPathParser::scanAttributeValue):
* Source/WebCore/layout/formattingContexts/inline/text/TextUtil.cpp:
(WebCore::Layout::TextUtil::containsStrongDirectionalityText):

Canonical link: https://commits.webkit.org/279356@main
